test/HcalDigiPipelineTest.cxx

```diff
@@ -0,0 +1,259 @@
+ */
+class HcalCheckEnergyReconstruction : public framework::Analyzer {
+
+public:
+  HcalCheckEnergyReconstruction(const std::string &name, framework::Process &p)
+      : framework::Analyzer(name, p) {}
+  ~HcalCheckEnergyReconstruction() {}
+
+  void onProcessStart() final override {
+
+    getHistoDirectory();
+    ntuple_.create("HcalDigiTest");
+    ntuple_.addVar<float>("HcalDigiTest", "SimEnergy");
+    ntuple_.addVar<float>("HcalDigiTest", "RecEnergy");
+    ntuple_.addVar<float>("HcalDigiTest","SimX");
+    ntuple_.addVar<float>("HcalDigiTest","SimY");
+    ntuple_.addVar<float>("HcalDigiTest","SimZ");
+    ntuple_.addVar<float>("HcalDigiTest","RecX");
+    ntuple_.addVar<float>("HcalDigiTest","RecY");
+    ntuple_.addVar<float>("HcalDigiTest","RecZ");
+    
+    ntuple_.addVar<int>("HcalDigiTest", "DaqDigi");
+    ntuple_.addVar<int>("HcalDigiTest", "DaqDigiIsADC");
+    ntuple_.addVar<int>("HcalDigiTest", "DaqDigiADC");
+    ntuple_.addVar<int>("HcalDigiTest", "DaqDigiTOT");
+  }
+
+  void analyze(const framework::Event &event) final override {
+
+    const auto simHits = event.getCollection<ldmx::SimCalorimeterHit>("HcalSimHits");
+
+    REQUIRE(simHits.size() == 1);
+
+    float truth_energy = simHits.at(0).getEdep();
+    ntuple_.setVar<float>("SimEnergy", truth_energy);
+    ntuple_.setVar<float>("SimX", simHits.at(0).getPosition()[0]);
+    ntuple_.setVar<float>("SimY", simHits.at(0).getPosition()[1]);
+    ntuple_.setVar<float>("SimZ", simHits.at(0).getPosition()[2]);
+    
+    const auto daqDigis{
+      event.getObject<ldmx::HgcrocDigiCollection>("HcalDigis")};
+    auto daqDigi = daqDigis.getDigi(0);
+    ntuple_.setVar<int>("DaqDigi",daqDigi.soi().raw());
+    bool is_in_adc_mode = daqDigi.isADC();
+    ntuple_.setVar<int>("DaqDigiIsADC",is_in_adc_mode);
+    ntuple_.setVar<int>("DaqDigiADC",daqDigi.soi().adc_t());
+    ntuple_.setVar<int>("DaqDigiTOT",daqDigi.tot());
+
+    const auto recHits = event.getCollection<ldmx::HcalHit>("HcalRecHits");
+    CHECK(recHits.size() == 1);
+
+    auto hit = recHits.at(0);
+    ldmx::HcalID id(hit.getID());
+    CHECK_FALSE(hit.isNoise());
+    CHECK(id.raw() == simHits.at(0).getID());
+
+    ntuple_.setVar<float>("RecX", hit.getXPos());
+    ntuple_.setVar<float>("RecY", hit.getYPos());
+    ntuple_.setVar<float>("RecZ", hit.getZPos());
+    
+    // define target energy by using the settings at the top
+    auto target_daq_energy = Approx(truth_energy).epsilon(MAX_ENERGY_PERCENT_ERROR_DAQ_TOT_MODE/100);
+    if (is_in_adc_mode) target_daq_energy = Approx(truth_energy).margin(MAX_ENERGY_ERROR_DAQ_ADC_MODE);
+
+    std::cout << "get ampl " << hit.getEnergy() << " taget daq " << truth_energy << " + " << MAX_ENERGY_ERROR_DAQ_ADC_MODE << std::endl;
+    CHECK(hit.getEnergy() == target_daq_energy);
+    ntuple_.setVar<float>("RecEnergy", hit.getEnergy());
+    
+    std::cout << "get x " << hit.getXPos() << " y " << hit.getYPos() << " z " << hit.getZPos() << std::endl;
```

Comment:
These messages are *super* helpful for developing, but comment them out before merging this in so that they don't pop up for someone who doesn't want them.